### PR TITLE
feature: extend website functionality

### PR DESCRIPTION
**Pull request by saenai255**
### Updates
  - [x] Rewrite client in `React`
  - [x] Create backend in `Go` 
  - [x] Implement package parsing in `Go`
  - [x] Create a package-list page
  - [x] Add search, sort, filter and pagination functionalities  
  - [x] Create build script
  - [x] Create build/release workflow that publishes a gzip containing standalone binaries
  - [x] Create feature flags to hide `PackageListPage`, `PackageDetailsPage` and to switch between the old and new cli syntax
  - [ ] Create package details page ( Only partially complete. It is currently hidden behind a feature flag ) 
  - [x] Dark mode

### Build Requirements
 - `node` ( version `16` or greater )
 - `go` ( version `1.17` or greater )
 - `make`

### Runtime Requirements
 - *None*

### How to Run
This will open a local http server at `http://localhost:3300`
```sh
make redist
# Rename ./redist/webpacd.toml.dist to ./redist/webpacd.toml
# Clone the pacstall-programs repository as specified in webpacd.toml
cd redist && ./webpacd
```

#### or if `webpacd.toml` and `pacstall-programs` are already set up
```sh
make run
```

### How to Build
```sh
make
```

### CODE REVIEW

File: client/src/pages/Home.tsx [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+				style={{ textAlign: 'center' }}>
+				<h3 className="uk-card-title" style={{ textAlign: 'center' }}>Why another package manager?</h3>
+				<p>If you've ever used the AUR, you know that every package known to man is there. But why is this Arch
+					specific? That's why I made Pacstall. An AUR-like alternative, for Ubuntu!</p>
+			</div>
+			<div className="uk-card uk-card-default uk-card-body uk-width-1-2@m uk-flex uk-flex-center uk-flex-wrap uk-card-hover uk-align-center"
+				style={{ textAlign: 'center' }}>
+				<h3 className="uk-card-title">How does it work then?</h3>
+				<p>Pacstall takes in files known as <a
+					href="https://github.com/pacstall/pacstall-programs/blob/master/make-a-pacscript.md">pacscripts</a> (similar to PKGBUILD's) that contain the necessary contents to build packages, and builds them into
```
Comment (Henryws) [inline client/src/pages/Home.tsx:24]:
```suggestion
					href="https://github.com/pacstall/pacstall/wiki/Pacscript-101</a> (similar to PKGBUILD's) that contain the necessary contents to build packages, and builds them into
```

**Review by Henryws — CHANGES_REQUESTED** [review]

**Comment by Henryws** [comment]
https://app.codacy.com/gh/pacstall/website/pullRequest?prid=8881665

Can you fix critical issues

File: .vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+    "configurations": [
+        {
+            "name": "Launch file",
+            "type": "go",
+            "request": "launch",
+            "mode": "debug",
+            "program": "server/main.go"
+        }
+    ]
+}
```
Comment (wizard-28) [inline .vscode/launch.json:15]: Sus.

File: .gitignore [hunk 3]
```diff
@@ -0,0 +1,2 @@
+redist/
+webpacd.tar.gz
```
Comment (wizard-28) [inline .gitignore:2]: Add `vscode` or better yet just use one of the standard gitignore templates for the project.

https://github.com/github/gitignore

File: client/src/index.tsx [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+import { BrowserRouter, Navigate, Route, Routes } from "react-router-dom";
+import { lazy, Suspense } from "react";
+import Home from "./pages/Home";
+import Showcase from "./pages/Showcase";
+import NotFound from "./pages/NotFound";
+import { QueryParamProvider } from "use-query-params";
+import axios from "axios";
+import { setupCache } from "axios-cache-adapter";
+import Notifications from "./components/Notifications";
+import { RecoilRoot } from "recoil";
```
Comment (wizard-28) [inline client/src/index.tsx:15]: Some lines have semi colons, others don't?

File: client/src/pages/Home.tsx [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+				<h3 className="uk-card-title">How does it work then?</h3>
+				<p>Pacstall takes in files known as <a
+					href="https://github.com/pacstall/pacstall/wiki/Pacscript-101</a> (similar to PKGBUILD's) that contain the necessary contents to build packages, and builds them into
+					executables on your system.</p>
+			</div>
+			<div className="uk-card uk-card-default uk-card-body uk-width-1-2@m uk-flex uk-flex-center uk-flex-wrap uk-card-hover uk-align-center"
+				style={{ textAlign: 'center' }}>
+				<h3 className="uk-card-title">Why is this any different than any other package manager?</h3>
+				<p>Pacstall uses the stable base of Ubuntu but allows you to use bleeding edge software with little to no
+					compromises, so you don't have to worry about security patches or new features.</p>
```
Comment (wizard-28) [inline client/src/pages/Home.tsx:31]: @Henryws what about this section, sounds like "will attempt to" to me

File: server/go.sum [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+github.com/BurntSushi/toml v1.0.0/go.mod h1:CxXYINrC8qIiEnFrOxCa7Jy5BFHlXnUU2pbicEuybxQ=
+github.com/google/uuid v1.3.0 h1:t6JiXgmwXMjEs8VusXIJk2BXHsn+wx8BZdTaoZ5fu7I=
+github.com/google/uuid v1.3.0/go.mod h1:TIyPZe4MgqvfeYDBFedMoGGpEw/LqOeaOT+nhxU+yHo=
+github.com/gorilla/mux v1.8.0 h1:i40aqfkR1h2SlN9hojwV5ZA91wcXFOvkdNIeFDP5koI=
+github.com/gorilla/mux v1.8.0/go.mod h1:DVbg23sWSpFRCP0SfiEN6jmj59UnW/n46BH5rLB71So=
+github.com/pelletier/go-toml v1.9.4 h1:tjENF6MfZAg8e4ZmZTeWaWiT2vXtsoO6+iuOjFhECwM=
+github.com/pelletier/go-toml v1.9.4/go.mod h1:u1nR/EPcESfeI/szUZKdtJ0xRNbUoANCkoOuaOx1Y+c=
+gopkg.in/check.v1 v0.0.0-20161208181325-20d25e280405/go.mod h1:Co6ibVJAznAaIkqp8huTwlJQCZ016jof/cbN4VW5Yz0=
+gopkg.in/yaml.v2 v2.4.0 h1:D8xgwECY7CYvx+Y2n4sBz93Jn9JRvxdiyyo8CTfuKaY=
+gopkg.in/yaml.v2 v2.4.0/go.mod h1:RDklbk79AGWmwhnvt/jBztapEOGDOx6ZbXqjP6csGnQ=
```
Comment (wizard-28) [inline server/go.sum:11]: Looks like a lock file to me... You sure this is meant to be merged?

**Review by wizard-28 — CHANGES_REQUESTED** [review]
Add new lines to the end of each file.

**Review by Henryws — CHANGES_REQUESTED** [review]
image

Should be able to copy on click

File: client/src/pages/Home.tsx (hunk 5 above)
Comment (wizard-28) [reply client/src/pages/Home.tsx:31]: To what

Comment (Henryws) [reply client/src/pages/Home.tsx:31]: I think we can move the `Why is this any different than any other package manager` to the top and remove the `Why another package manager`

File: server/go.sum (hunk 6 above)
Comment (saenai255) [reply server/go.sum:11]: Lock files are intended to be merged

**Comment by saenai255** [comment]
> image
> 
> Should be able to copy on click

Last time I tested, it worked. Will look into it again

**Comment by saenai255** [comment]
Added dark mode. [See commit](https://github.com/pacstall/website/pull/19/commits/111973323698dfc27b4013700e4851071741fbcb)

File: client/public/styles/global.css [hunk 7]
```diff
@@ -0,0 +1 @@
+@import url('https://fonts.googleapis.com/css2?family=Source+Code+Pro:wght@600&display=swap');
```
Comment (Henryws) [inline client/public/styles/global.css:1]: I would prefer if we hosted the font locally for speed, and also because I don't like google.

**Review by Henryws — CHANGES_REQUESTED** [review]

**Review by Henryws — CHANGES_REQUESTED** [review]
Running `make redist` results in this error
```
henry@twilight ~/pacstall-website % ››› make redist                                   [master]
which node
/usr/bin/node
npm i -g yarn
npm ERR! code EACCES
npm ERR! syscall mkdir
npm ERR! path /usr/lib/node_modules/yarn
npm ERR! errno -13
npm ERR! Error: EACCES: permission denied, mkdir '/usr/lib/node_modules/yarn'
npm ERR!  [Error: EACCES: permission denied, mkdir '/usr/lib/node_modules/yarn'] {
npm ERR!   errno: -13,
npm ERR!   code: 'EACCES',
npm ERR!   syscall: 'mkdir',
npm ERR!   path: '/usr/lib/node_modules/yarn'
npm ERR! }
npm ERR!
npm ERR! The operation was rejected by your operating system.
npm ERR! It is likely you do not have the permissions to access this file as the current user
npm ERR!
npm ERR! If you believe this might be a permissions issue, please double-check the
npm ERR! permissions of the file and its containing directories, or try running
npm ERR! the command again as root/Administrator.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/henry/.npm/_logs/2022-02-27T16_13_42_171Z-debug-0.log
make: *** [makefile:16: client/dist] Error 243
```

**Comment by Henryws** [comment]
image
Copying clips into other text

**Review by Henryws — CHANGES_REQUESTED** [review]
image
Make the `$` a different color

**Review by Henryws — CHANGES_REQUESTED** [review]
image
Double `ff` looks weird

File: client/src/pages/Home.tsx [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+// @ts-ignore:next-line
+import PacstallLogo from '../../public/pacstall.svg'
+
+
+const Home: FC = () => {
+	return (<>
+		<Navigation />
+		<Container maxW='900px'>
+
+			<Heading size='2xl' textAlign='center' p='7'>Pacstall - The AUR for Ubuntu</Heading>
```
Comment (Henryws) [inline client/src/pages/Home.tsx:15]:
```suggestion
			<Heading size='2xl' textAlign='center' p='7'>Pacstall ~ The AUR for Ubuntu</Heading>
```

**Review by Henryws — COMMENTED** [review]
@wizard-28 your thoughts on this? (~ instead of -)
image

File: makefile [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+	[ -d ./redist/pacstall-programs ] && rm -rf redist/pacstall-programs || :
+	git clone https://github.com/pacstall/pacstall-programs redist/pacstall-programs
+	cp -r client/dist/* redist/public
+	cp -r server/bin/* redist
+
+
+#### Commands
+.PHONY: run clean
+
+run: redist run-dist
```
Comment (Henryws) [inline makefile:30]:
```suggestion
run:
```

**Review by Henryws — CHANGES_REQUESTED** [review]

**Review by Henryws — CHANGES_REQUESTED** [review]
```systemd
[Unit]
Description=Pacstall Web Server

[Service]
WorkingDirectory=/var/www/pacstall/
ExecStart=/var/www/pacstall/webpacd
Restart=always

[Install]
WantedBy=multi-user.target
WantedBy=network-online.target
```

Add a systemd script like this

**Review by Henryws — CHANGES_REQUESTED** [review]
image
Swap `Social Networks` and `Showcase`

**Review by Henryws — CHANGES_REQUESTED** [review]
image
Second section mentions opening a terminal, but that should be done in step one

**Review by Henryws — CHANGES_REQUESTED** [review]
Clicking search twice with no input goes in an infinite loop

**Review by Henryws — CHANGES_REQUESTED** [review]
image
`git` is listed as a pacdep when it's only a build dep

File: client/src/pages/PrivacyPolicy.tsx [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+            <Text>We have no control over and assume no responsibility for the content, privacy policies or practices of any third party sites or services.</Text>
+            <Heading mb='3' mt='7'>Changes to this Privacy Policy</Heading>
+            <Text>We may update Our Privacy Policy from time to time. We will notify You of any changes by posting the new Privacy Policy on this page.</Text>
+            <Text>We will let You know via email and/or a prominent notice on Our Service, prior to the change becoming effective and update the &quot;Last updated&quot; date at the top of this Privacy Policy.</Text>
+            <Text>You are advised to review this Privacy Policy periodically for any changes. Changes to this Privacy Policy are effective when they are posted on this page.</Text>
+            <Heading mb='3' mt='7'>Contact Us</Heading>
+            <Text>If you have any questions about this Privacy Policy, You can contact us:</Text>
+            <UnorderedList>
+                <ListItem>By email: pacstall@protonmail.com</ListItem>
```
Comment (Henryws) [inline client/src/pages/PrivacyPolicy.tsx:141]: This should be clickable

**Review by Henryws — CHANGES_REQUESTED** [review]

**Review by Henryws — CHANGES_REQUESTED** [review]
- [ ] [back button sometimes works](https://discord.com/channels/839818021207801878/889128266416422953/949676607256272916)

- [ ] [git is showing as a pacdeps for emacs, even though it is listed as a build depends](https://discord.com/channels/839818021207801878/889128266416422953/949678207513268254)

- [ ] [useless search animation](https://discord.com/channels/839818021207801878/889128266416422953/949678765447987200)

- [ ] searching for a package loads the page twice

- [ ] `Package Repository` is listed under Downloads

- [ ]  ipad orentation header

- [ ] iphone 11 header

- [ ] slow animations for notifications on copy in brave

- [ ] if variables are separated by a `\`, the API will return `""` for every space

- look at stellarium

- [ ] packageName in API is empty if `pkgname` is non-existent (replace with name if pkgname doesn't exist)

- appimagelauncher-deb

- [ ] tab titles don't change

- [ ] if you visit the site for the first time and you visit the privacy page first, going anywhere else bypasses the cookie prompt

- [ ] menu bar buttons don't work on landscape mode

- [x] showcase page is unusable on phone

- [x] long gap between `interpretation` and `and` (privacy page) when using phone size

- [ ] `how does it work then` paragraph needs to be rephrased

---

Notes taken when @wizard-28  was having a bug fest

**Comment by wizard-28** [comment]
@Henryws you didn't add all of them to the list, example this one: https://discord.com/channels/839818021207801878/889128266416422953/949735716873723935

**Comment by wizard-28** [comment]
Let's gooooo.

**Pull request merged**